src/polyfills/fetch.js

```diff
@@ -0,0 +1,307 @@
+   * @return {!Promise<!JsonObject>}
+   */
+  json() {
+    return /** @type {!Promise<!JsonObject>} */ (
+      this.drainText_().then(parseJson));
+  }
+
+  /**
+   * Drains the response and returns a promise that resolves with the response
+   * ArrayBuffer.
+   * @return {!Promise<!ArrayBuffer>}
+   */
+  arrayBuffer() {
+    return /** @type {!Promise<!ArrayBuffer>} */ (
+      this.drainText_().then(utf8Encode));
+  }
+}
+
+/**
+ * Normalized method name by uppercasing.
+ * @param {string|undefined} method
+ * @return {string}
+ * @private
+ */
+function normalizeMethod(method) {
+  if (method === undefined) {
+    return 'GET';
+  }
+  method = method.toUpperCase();
+  dev().assert(
+      allowedMethods.includes(method),
+      'Only one of %s is currently allowed. Got %s',
+      allowedMethods.join(', '),
+      method
+  );
+  return method;
+}
+
+
+/**
+ * Provides access to the response headers as defined in the Fetch API.
+ * @private Visible for testing.
+ */
+class FetchResponseHeaders {
+  /**
+   * @param {!XMLHttpRequest|!XDomainRequest|!XMLHttpRequestDef} xhr
+   */
+  constructor(xhr) {
+    /** @private @const {!XMLHttpRequest|!XDomainRequest|!XMLHttpRequestDef} */
+    this.xhr_ = xhr;
+  }
+
+  /**
+   * @param {string} name
+   * @return {string}
+   */
+  get(name) {
+    return this.xhr_.getResponseHeader(name);
+  }
+
+  /**
+   * @param {string} name
+   * @return {boolean}
+   */
+  has(name) {
+    return this.xhr_.getResponseHeader(name) != null;
+  }
+}
+/**
+ * Returns instance of Response.
+ * @param {?ResponseBodyInit=} body
+ * @param {!ResponseInit=} init
+ */
+export function Response(body, init = {}) {
+  const lowercasedHeaders = map();
+  const data = Object.assign({
+    status: 200,
+    statusText: 'OK',
+    responseText: (body ? String(body) : ''),
+    /**
+     * @param {string} name
+     * @return {string}
+     */
+    getResponseHeader(name) {
+      return lowercasedHeaders[String(name).toLowerCase()] || null;
+    },
+  }, init);
+
+  init.status = init.status || 200;
+
+  if (isArray(init.headers)) {
+    init.headers.forEach(entry => {
+      const headerName = entry[0];
+      const headerValue = entry[1];
+      lowercasedHeaders[String(headerName).toLowerCase()] =
+          String(headerValue);
+    });
+  } else if (isObject(init.headers)) {
+    for (const key in init.headers) {
+      lowercasedHeaders[String(key).toLowerCase()] =
+          String(init.headers[key]);
+    }
+  }
+  if (init.status) {
+    data.status = parseInt(init.status, 10);
+  }
+  if (init.statusText) {
+    data.statusText = String(init.statusText);
+  }
+
+  return new FetchResponse(/** @type {XMLHttpRequestDef} */(data));
+}
+
+/**
+ * Installs fetch and Response polyfill
+ *
+ * @export
+ * @param {Window} win
+ */
+export function install(win) {
+  if (!win.fetch) {
+    win.fetch = /** @type {function(string, RequestInit):!Promise} */ (fetchPolyfill);
+    win.Response = Response;
```

Comment:
Use `Object.defineProperty`.

Reply:
```
Object.getOwnPropertyDescriptor(window, 'Response')
{value: ƒ, writable: true, enumerable: false, configurable: true}
Object.getOwnPropertyDescriptor(window, 'fetch')
{value: ƒ, writable: true, enumerable: true, configurable: true}
```

Enumerability, in particular, has been a source of countless polyfill bugs.

Reply:
done

Reply:
I am curious, can u list one such example?

Reply:
https://developers.google.com/web/updates/2018/03/smooshgate, https://bugzilla.mozilla.org/show_bug.cgi?id=781796.

Reply:
Gotcha! I have read that before, but now I got the complete understanding. Thanks